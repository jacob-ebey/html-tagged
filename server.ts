export type {
  CustomElement,
  CustomElementArgs,
  RenderToStringArgs,
  RenderToStringContext,
} from "./lib/render.ts";
export { renderToString } from "./lib/render.ts";
