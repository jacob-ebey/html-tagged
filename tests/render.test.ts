import { assertSnapshot } from "snapshot";

import { html, value } from "../mod.ts";
import { renderToString } from "../server.ts";

Deno.test("should render a simple string", async (t) => {
  const result = renderToString(html`<div></div>`);
  await assertSnapshot(t, result);
});

Deno.test(
  "should render custom element without implementation to string",
  async (t) => {
    const result = renderToString(
      html`<div><custom-element></custom-element></div>`,
    );
    await assertSnapshot(t, result);
  },
);

Deno.test(
  "should render nested custom elements without implementation to string",
  async (t) => {
    const result = renderToString(
      html`<div>
        <custom-element-a>
          a
          <custom-element-b>b</custom-element-b>
        </custom-element-a>
      </div>`,
    );
    await assertSnapshot(t, result);
  },
);

Deno.test("should expand custom element with implementation", async (t) => {
  const result = renderToString(
    html`<div><custom-element></custom-element></div>`,
    {
      elements: {
        "custom-element": () => html`<div>custom element</div>`,
      },
    },
  );
  await assertSnapshot(t, result);
});

Deno.test("should expand custom element with slot", async (t) => {
  const result = renderToString(
    html`<div>
      <custom-element><span>a</span></custom-element>
    </div>`,
    {
      elements: {
        "custom-element": () => html`<div>custom element <slot></slot></div>`,
      },
    },
  );
  await assertSnapshot(t, result);
});

Deno.test("should pass attrs to custom-element", async (t) => {
  const result = renderToString(
    html`<div>
      <custom-element name="test"><span>a</span></custom-element>
    </div>`,
    {
      elements: {
        "custom-element": ({ attrs }) =>
          html`<div>custom element ${value(attrs.name)} <slot></slot></div>`,
      },
    },
  );
  await assertSnapshot(t, result);
});

Deno.test("should expand nested custom element with slot", async (t) => {
  const result = renderToString(
    html`<div>
      <custom-element>
        <span>a</span>
        <custom-element>
          <span>b</span>
        </custom-element>
      </custom-element>
    </div>`,
    {
      elements: {
        "custom-element": () => html`<div>custom element <slot></slot></div>`,
      },
    },
  );
  await assertSnapshot(t, result);
});

// TODO: Allow expansion of custom elements within a custom elements template.
// it.skip("should expand custom element in another custom element implementation", async (t) => {
//   const result = renderToString(
//     html`<div>
//       <custom-element-a><span>outer</span></custom-element-a>
//     </div>`,
//     {
//       elements: {
//         "custom-element-a": () =>
//           html`a<custom-element-b><slot></slot></custom-element-b>`,
//         "custom-element-b": () => html`<div>b <slot></slot></div>`,
//       },
//     }
//   );
//   inlineSnapshot(
//     result,
//     "<div>\n        <custom-element-a>a<custom-element-b><div>b <span>outer</span></div></custom-element-b></custom-element-a>\n      </div>"
//   );
// });

Deno.test("should render style in place if no body in tree", async (t) => {
  const result = renderToString(
    html`<p>a</p>
      <style>
        html {
          color: green;
        }
      </style>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should lift styles to head", async (t) => {
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            <style>
              html {
                color: green;
              }
            </style>
            <script>
              console.log("hello");
            </script>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should lift styles to head for sub HTMLNode", async (t) => {
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            ${html`<style>
              html {
                color: green;
              }
            </style>`}
            <script>
              console.log("hello");
            </script>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should lift styles to head and dedupe", async (t) => {
  const sub = html`<style>
    html {
      color: green;
    }
  </style>`;
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            ${sub} ${sub}
            <script>
              console.log("hello");
            </script>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should render scripts in place if no body in tree", async (t) => {
  const result = renderToString(
    html`<p>a</p>
      <script>
        console.log("hello");
      </script>`,
  );
  await assertSnapshot(t, result);
});

Deno.test(
  "should render nested scripts in place if no body in tree",
  async (t) => {
    const result = renderToString(
      html`<script>
          console.log("a");
        </script>
        ${html`<script>
          console.log("b");
        </script>`}`,
    );
    await assertSnapshot(t, result);
  },
);

Deno.test("should put scripts at end of body", async (t) => {
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            <style>
              html {
                color: green;
              }
            </style>
            <script>
              console.log("hello");
            </script>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should put scripts at end of body for sub HTMLNode", async (t) => {
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            ${html`<script>
              console.log("hello");
            </script>`}
            <style>
              html {
                color: green;
              }
            </style>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});

Deno.test("should put scripts at end of body and dedupe", async (t) => {
  const sub = html`<script>
    console.log("hello");
  </script>`;
  const result = renderToString(
    html`<!DOCTYPE html>
      <html>
        <head></head>
        <body>
          <div>
            ${sub} ${sub}
            <style>
              html {
                color: green;
              }
            </style>
          </div>
        </body>
      </html>`,
  );
  await assertSnapshot(t, result);
});
