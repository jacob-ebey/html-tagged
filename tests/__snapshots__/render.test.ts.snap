export const snapshot = {};

snapshot[`should render a simple string 1`] = `"<div></div>"`;

snapshot[`should render custom element without implementation to string 1`] = `"<div><custom-element></custom-element></div>"`;

snapshot[`should render nested custom elements without implementation to string 1`] = `
"<div>
" +
  "        <custom-element-a>
" +
  "          a
" +
  "          <custom-element-b>b</custom-element-b>
" +
  "        </custom-element-a>
" +
  "      </div>"
`;

snapshot[`should expand custom element with implementation 1`] = `"<div><custom-element><div>custom element</div></custom-element></div>"`;

snapshot[`should expand custom element with slot 1`] = `
"<div>
" +
  "      <custom-element><div>custom element <span>a</span></div></custom-element>
" +
  "    </div>"
`;

snapshot[`should pass attrs to custom-element 1`] = `
"<div>
" +
  '      <custom-element name="test"><div>custom element test <span>a</span></div></custom-element>
' +
  "    </div>"
`;

snapshot[`should expand nested custom element with slot 1`] = `
"<div>
" +
  "      <custom-element><div>custom element 
" +
  "        <span>a</span>
" +
  "        <custom-element><div>custom element 
" +
  "          <span>b</span>
" +
  "        </div></custom-element>
" +
  "      </div></custom-element>
" +
  "    </div>"
`;

snapshot[`should render style in place if no body in tree 1`] = `
"<p>a</p>
" +
  "      <style>
" +
  "        html {
" +
  "          color: green;
" +
  "        }
" +
  "      </style>"
`;

snapshot[`should lift styles to head 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "              html {
" +
  "                color: green;
" +
  "              }
" +
  "            </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "            
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '              console.log("hello");
' +
  "            </script></body>
" +
  "      </html>"
`;

snapshot[`should lift styles to head for sub HTMLNode 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "              html {
" +
  "                color: green;
" +
  "              }
" +
  "            </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "            
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '              console.log("hello");
' +
  "            </script></body>
" +
  "      </html>"
`;

snapshot[`should lift styles to head and dedupe 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "    html {
" +
  "      color: green;
" +
  "    }
" +
  "  </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "             
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '              console.log("hello");
' +
  "            </script></body>
" +
  "      </html>"
`;

snapshot[`should render scripts in place if no body in tree 1`] = `
'<p>a</p>
      <script>
        console.log("hello");
      </script>'
`;

snapshot[`should render nested scripts in place if no body in tree 1`] = `
"<script>
" +
  '          console.log("a");
' +
  "        </script>
" +
  "        <script>
" +
  '          console.log("b");
' +
  "        </script>"
`;

snapshot[`should put scripts at end of body 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "              html {
" +
  "                color: green;
" +
  "              }
" +
  "            </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "            
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '              console.log("hello");
' +
  "            </script></body>
" +
  "      </html>"
`;

snapshot[`should put scripts at end of body for sub HTMLNode 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "              html {
" +
  "                color: green;
" +
  "              }
" +
  "            </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "            
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '              console.log("hello");
' +
  "            </script></body>
" +
  "      </html>"
`;

snapshot[`should put scripts at end of body and dedupe 1`] = `
"<!DOCTYPE html>
" +
  "      <html>
" +
  "        <head><style>
" +
  "              html {
" +
  "                color: green;
" +
  "              }
" +
  "            </style></head>
" +
  "        <body>
" +
  "          <div>
" +
  "             
" +
  "            
" +
  "          </div>
" +
  "        <script>
" +
  '    console.log("hello");
' +
  "  </script></body>
" +
  "      </html>"
`;
